### PR TITLE
Fixing byte compilation issues

### PR DESCRIPTION
**Pull request by phikal**
Hi,

I would like to add gnu-apl-mode to [NonGNU ELPA](http://elpa.nongnu.org/), but before that is done, a few code issues should be fixed. The attached patches improve remove the deprecated cl library, and fix a number of issues raised by the byte compiler (see commit description).

If you decide to apply these changes, it would be necessary to bump the [version attribute in the main file](https://github.com/lokedhs/gnu-apl-mode/blob/master/gnu-apl-mode.el#L6), so that ELPA recognizes a new release with these improvements.

### CODE REVIEW

**Comment by lokedhs** [comment]
Thanks. This sounds like a good idea. At least until I can get the copyright assignments done so that it can be posted to ELPA.

I'll take a look at your proposed changes.

**Comment by phikal** [comment]
ELPA is always preferable, but I am not sure what the state if of the other contributors with significant contributions (@fourier, @ doyougnu, @ecraven). It might be manageable, but if for whatever reason it wouldn't be, I think it would be preferable to just add the package to NonGNU ELPA, that doesn't require any CA.

**Comment by lokedhs** [comment]
Yes, that's why I think it's a good idea. I have no idea when the copyright assignments will be done.

**Comment by fourier** [comment]
I have signed copyright papers and have code in ELPA if it helps.

**Comment by lokedhs** [comment]
I'd prefer not to merge the changes as-is, since they change spaces to tabs. Can you recommit with only spaces?

The commits themselves looks find to me, but I might have missed something due to the noise caused by the spaces being changed to tabs.

**Comment by phikal** [comment]
Sorry about that, should be fixed now. I also hope it is ok to add a .dir-locals.el to make sure that this doesn't happen again in the future.

File: gnu-apl-editor.el [hunk 1]
```diff
@@ -159,7 +169,8 @@ successfully."
         (gnu-apl--send-new-function parts tag)))))
 
 (defun gnu-apl--remove-final-endfn (strings)
-  "If the last element is ∇, return a new list with that element removed, else return the original list."
+  "Remove ∇ if it is the last element in STRINGS.
```
Comment (lokedhs) [inline gnu-apl-editor.el:172]: While I generally don't have a problem with rewrites of docstrings, this change makes it seem as though the string is modified in-place, which is not the case.

File: gnu-apl-follow.el [hunk 2]
```diff
@@ -4,18 +4,24 @@
 (require 'gnu-apl-util)
 (require 'gnu-apl-network)
 
+(declare-function gnu-apl--get-interactive-session "gnu-apl-interactive")
+(declare-function gnu-apl--name-at-point "gnu-apt-documentation")
+(declare-function gnu-apl--choose-variable "gnu-apl-editor"
+                  (prompt &optional type default-value))
+(declare-function gnu-apl--get-interactive-session-with-nocheck "gnu-apt-interactive")
```
Comment (lokedhs) [inline gnu-apl-follow.el:11]: Is `gnu-apt-interactive` a typo?

File: gnu-apl-follow.el [hunk 3]
```diff
@@ -4,18 +4,24 @@
...
+(defvar gnu-apl-trace-variable)
+
+(defvar gnu-apl-trace-symbols nil
+  "List of traced symbols.
+Each element has the structure (\"symbol_name\" <buffer>).")
 
 (defun gnu-apl-trace-mode-kill-buffer ()
-  "If the current buffer is a trace buffer, kill the buffer. Otherwise raise an error."
+  "If the current buffer is a trace buffer, kill the buffer.
+Otherwise raise an error."
```
Comment (lokedhs) [inline gnu-apl-follow.el:24]: What is the reason for a newline here? It breaks the flow of the text and when only the first line is displayed, not all information is shown.

File: gnu-apl-interactive.el [hunk 4]
```diff
@@ -117,9 +168,7 @@ code was read from."
 (defun gnu-apl--erase-and-set-function (name _content)
   (gnu-apl-interactive-send-string (concat "'" *gnu-apl-ignore-start* "'"))
   (gnu-apl-interactive-send-string (concat ")ERASE " name))
-  (gnu-apl-interactive-send-string (concat "'" *gnu-apl-ignore-end* "'"))
-  (setq gnu-apl-function-content-lines (split-string content "\n"))
-  (gnu-apl-interactive-send-string (concat "'" *gnu-apl-send-content-start* "'")))
+  (gnu-apl-interactive-send-string (concat "'" *gnu-apl-ignore-end* "'")))
```
Comment (lokedhs) [inline gnu-apl-interactive.el:171]: This seems to be a mistake. The code that updates the function is gone.

File: gnu-apl-network.el [hunk 5]
```diff
@@ -43,8 +58,7 @@ connect mode in use."
 
 (defun gnu-apl--connect (connect-mode addr)
   (with-current-buffer (gnu-apl--get-interactive-session)
-    (when (and (boundp 'gnu-apl--connection)
-               (process-live-p gnu-apl--connection))
+    (when (process-live-p gnu-apl--connection)
```
Comment (lokedhs) [inline gnu-apl-network.el:61]: What is the reason for this change? The binding of `gnu-apl--connection` in the buffer only was intentional. Granted, the behaviour should be the same, but it's nice not to pollute the namespace with symbols that can never be used outside an apl buffer.

File: gnu-apl-osx-workaround.el [hunk 6]
```diff
@@ -3,6 +3,8 @@
 (require 'cl-lib)
 (require 'gnu-apl-util)
 
+(defvar gnu-apl--symbols)                   ;gnu-apl-symbols.el
```
Comment (lokedhs) [inline gnu-apl-osx-workaround.el:6]: Why not just `require` here?

File: gnu-apl-plot.el [hunk 7]
```diff
@@ -4,6 +4,12 @@
 (require 'gnu-apl-util)
 (require 'gnu-apl-network)
 
+(declare-function gnu-apl--name-at-point "gnu-apl-documentation" ())
+(declare-function gnu-apl--choose-variable "gnu-apl-editor"
+		  (prompt &optional type default-value))
+
+(defvar gnu-apl-gnuplot-program)	;gnu-apl-mode.el
```
Comment (lokedhs) [inline gnu-apl-plot.el:11]: We can just `require` here.

File: gnu-apl-spreadsheet.el [hunk 8]
```diff
@@ -123,6 +129,9 @@ value of VARNAME to the content of the spreadsheet."
 values as the spreadsheet in the current buffer."
   (concat (gnu-apl-make-array-loading-instructions varname)))
 
+(defvar ses--numrows)                   ;ses.el
```
Comment (lokedhs) [inline gnu-apl-spreadsheet.el:132]: I'm not at all a fan of defining variables with names from other packages. This should not be needed.

**Review by lokedhs — CHANGES_REQUESTED** [review]
Thanks for your changes. I've added some notes.

File: gnu-apl-follow.el (hunk 2 above)
Comment (phikal) [reply gnu-apl-follow.el:11]: Yes, will fix.

File: gnu-apl-follow.el (hunk 3 above)
Comment (phikal) [reply gnu-apl-follow.el:24]: The reason is that checkdoc complains about documentation strings that are longer than 80 characters. If you want to, I can just ignore checkdoc for now.

File: gnu-apl-interactive.el (hunk 4 above)
Comment (phikal) [reply gnu-apl-interactive.el:171]: That might be the case, but I don't see where `*gnu-apl-send-content-start* ` is used.

File: gnu-apl-osx-workaround.el (hunk 6 above)
Comment (phikal) [reply gnu-apl-osx-workaround.el:6]: To be honest, I didn't want to worry about circular dependencies, and because most of the code was required by `gnu-apl-mode.el`, I just declared the variables this way.

File: gnu-apl-spreadsheet.el (hunk 8 above)
Comment (phikal) [reply gnu-apl-spreadsheet.el:132]: Without these declarations, I seem to get the following issues:

```
emacs -Q --batch -L . -f batch-byte-compile gnu-apl-spreadsheet.el

In gnu-apl-make-array-loading-instructions:
gnu-apl-spreadsheet.el:139:17: Warning: reference to free variable
    ‘ses--numrows’
gnu-apl-spreadsheet.el:140:17: Warning: reference to free variable
    ‘ses--numcols’
gnu-apl-spreadsheet.el:152:42: Warning: reference to free variable
    ‘ses--cells’
```

File: gnu-apl-editor.el (hunk 1 above)
Comment (phikal) [reply gnu-apl-editor.el:172]: How about "Remove ∇ non-destructively ..."?

**Comment by phikal** [comment]
If you are interested, I could also add the [Makefile](https://0x0.st/-YIc.bin) I was using to the project. It could be used to set up a CI (though I don't know how GitHub does that).

File: gnu-apl-editor.el (hunk 1 above)
Comment (lokedhs) [reply gnu-apl-editor.el:172]: I suggest: `Return STRINGS without the terminating ∇, if present`

File: gnu-apl-follow.el (hunk 3 above)
Comment (lokedhs) [reply gnu-apl-follow.el:24]: Yes, please do so in this case.

File: gnu-apl-interactive.el (hunk 4 above)
Comment (lokedhs) [reply gnu-apl-interactive.el:171]: I wrote that code a long time ago, and I believe that the entire function isn't needed anymore. Its job is now handled by `gnu-apl--send-si-and-send-new-function`.

File: gnu-apl-network.el (hunk 5 above)
Comment (lokedhs) [reply gnu-apl-network.el:61]: Do you have details?

File: gnu-apl-osx-workaround.el (hunk 6 above)
Comment (lokedhs) [reply gnu-apl-osx-workaround.el:6]: Unless I'm mistaken, there should be no circular dependency in this case.

File: gnu-apl-spreadsheet.el (hunk 8 above)
Comment (lokedhs) [reply gnu-apl-spreadsheet.el:132]: That's because these variables are buffer-local to SES buffers. They shouldn't be visible globally.

File: gnu-apl-interactive.el (hunk 4 above)
Comment (phikal) [reply gnu-apl-interactive.el:171]: You are right, the function is dead code. I'll remove it in that case.

File: gnu-apl-plot.el (hunk 7 above)
Comment (phikal) [reply gnu-apl-plot.el:11]: When you do that, there is a recursive require chain between gnu-apl-mode and gnu-apl-plot.

File: gnu-apl-osx-workaround.el (hunk 6 above)
Comment (phikal) [reply gnu-apl-osx-workaround.el:6]: You are right, that was overcaution on my side.

File: gnu-apl-spreadsheet.el (hunk 8 above)
Comment (phikal) [reply gnu-apl-spreadsheet.el:132]: But in this case they have to be "visible" as gnu-apl-spreadsheet.el isn't part of SES. After all, the `--` indicates that the variables are not actually part of the interface that SES provides. They have to at least be declare for the byte compiler, which doesn't mean they are bound, so I don't think this is a real issue.

File: gnu-apl-follow.el (hunk 3 above)
Comment (phikal) [reply gnu-apl-follow.el:24]: If you insist, but I'll just point out that this is the last remaining byte compilation issue.

File: gnu-apl-editor.el (hunk 1 above)
Comment (phikal) [reply gnu-apl-editor.el:172]: Done.

File: gnu-apl-network.el (hunk 5 above)
Comment (phikal) [reply gnu-apl-network.el:61]: It turns out that declaring `gnu-apl--connection` is enough to solve the issue, in which case the boundp isn't necessary. Without declaring the variable, you get a number of byte compilation issues along the lines of "Warning: reference to free variable ...".

File: gnu-apl-follow.el (hunk 3 above)
Comment (lokedhs) [reply gnu-apl-follow.el:24]: If you can come up with a better wording, I'm all for it.

Just so you see my concern: The issue is that the entire behaviour right now is concisely covered by a single sentence. Having an arbitrary 80 character limit break that is not ideal.

File: gnu-apl-network.el (hunk 5 above)
Comment (lokedhs) [reply gnu-apl-network.el:61]: Sure, but then we end up with the symbol being leaked outside the buffers where they belong. I blame the compiler here.

That said, as long as it works, it doesn't really matter, so just go ahead.

File: gnu-apl-plot.el (hunk 7 above)
Comment (lokedhs) [reply gnu-apl-plot.el:11]: Then I suggest we move the `defcustom` to `gnu-apl-plot.el` instead.

File: gnu-apl-follow.el (hunk 3 above)
Comment (phikal) [reply gnu-apl-follow.el:24]: Thinking about this some more, I tried setting `fill-collumn` in .dir-locals.el to a value greater than 80, and that seems to have "fixed" the issue. I'll have to check if ELPA accepts this too, but it seems like a first approximation of a satisfactory solution.

File: gnu-apl-spreadsheet.el (hunk 8 above)
Comment (lokedhs) [reply gnu-apl-spreadsheet.el:132]: True. Having to access internal symbols is not great. An issue is that SES doesn't really have an external API. Also, I'm pretty sure that when SES is compiled, you're going to see the same warnings.

Doing a `defvar` on another package's symbol is still bad, and while extremely unlikely, could still cause problems outside of gnu-apl-mode.

If you really want to get rid of that warning, one can always work around it using `eval`. Horrible, I know, but better than messing with another package's symbols.

File: gnu-apl-plot.el (hunk 7 above)
Comment (phikal) [reply gnu-apl-plot.el:11]: Done!

File: gnu-apl-spreadsheet.el (hunk 8 above)
Comment (phikal) [reply gnu-apl-spreadsheet.el:132]: I hesitate to use eval, because I don't think that the defvar approach is _that_ bad, considering that these are just declarations and not definitions. But if you insist, I'll make the change.

> Also, I'm pretty sure that when SES is compiled, you're going to see the same warnings.

It doesn't seem to raise any issues on my system (using a recent checkout of 28.0.60).

Comment (lokedhs) [reply gnu-apl-spreadsheet.el:132]: Interesting. Because you had to add these declarations in order to get rid of these warnings, but SES doesn't have them, yet don't have the warnings. Why is it different? I'd need to look at what SES is doing.

**Comment by phikal** [comment]
Ok, as you see I have added the Makefile to this push request. I have no experience with GitHub Builds, so I hope you can take care of that.

File: gnu-apl-spreadsheet.el (hunk 8 above)
Comment (lokedhs) [reply gnu-apl-spreadsheet.el:132]: OK, I'm frustrated by this one. I will merge your changes now, and I will at some point try to find out what is going on so the declaration can be avoided

**Comment by phikal** [comment]
Thank you for merging the changes.

In case you are still interested: All that is left for NonGNU ELPA would be to update the version tag, as indicated in the first message.
